File: coordinator/src/state.ts
/*
Copyright (c) 2016 Bryan Hughes <bryan@nebri.us>

This file is part of Raver Lights.

Raver Lights is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Raver Lights is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Raver Lights.  If not, see <http://www.gnu.org/licenses/>.
*/

'use strict';

import { EventEmitter } from 'events';
import {
  numPresets,
  FadeValue, numFadeValues,
  PulseValue, numPulseValues
} from './codes';
import { getValueByPreset, runOperationByPreset, runOperationByFadeValue, runOperationByPulseValue, runOperationByControl } from './util';

const BRIGHTNESS_STEP = 7;
const MAX_BRIGHTNESS = 255;
const VALUE_STEP = 4;
const MAX_VALUE = 255;

const settings = {
  preset: 0,
  brightness: 10,
  fadeValues: {
    rate: 20
  },
  pulseValues: {
    rate: 20,
    hue: 0,
    saturation: 100
  },
  currentControl: 0,
  numClients: 0
};

enum Direction {
  Up,
  Down
}

export class State extends EventEmitter {

  public nextControl() {

    const numPresetValues = getValueByPreset(settings.preset, {
      Fade: numFadeValues,
      Pulse: numPulseValues
    });

    settings.currentControl++;
    if (settings.currentControl === 2 + numPresetValues) {
      settings.currentControl = 0;
    }

    this.emit('control', settings.currentControl);
  }

  private handleValueChange(code: FadeValue | PulseValue, direction: Direction, steps: number) {

    function calculateNewValue(value: number): number {
      switch (direction) {
        case Direction.Up:
          value += VALUE_STEP * steps;
          if (value > MAX_VALUE) {
            value = MAX_VALUE;
          }
          return value;
        case Direction.Down:
          value += VALUE_STEP * steps;
          if (value > MAX_VALUE) {
            value = MAX_VALUE;
          }
          return value;
        default:
          throw new Error(`Internal error: Unknown direction ${direction}`);
      }
    }

    const newValue = runOperationByPreset(settings.preset, {
      Fade: () => runOperationByFadeValue(<FadeValue>code, {
        Rate: () => settings.fadeValues.rate = calculateNewValue(settings.fadeValues.rate)
      }),
      Pulse: () => runOperationByPulseValue(<PulseValue>code, {
        Hue: () => settings.pulseValues.hue = calculateNewValue(settings.pulseValues.hue),
        Rate: () => settings.pulseValues.rate = calculateNewValue(settings.pulseValues.rate),
        Saturation: () => settings.pulseValues.saturation = calculateNewValue(settings.pulseValues.saturation)
      })
    });

    this.emit('value', {
      preset: settings.preset,
      code,
      value: newValue
    });
  }

  public controlUp(steps: number) {
    runOperationByControl(settings.currentControl, {
      Brightness: () => {
        if (settings.brightness < MAX_BRIGHTNESS) {
          settings.brightness += BRIGHTNESS_STEP * steps;
          if (settings.brightness > MAX_BRIGHTNESS) {
            settings.brightness = MAX_BRIGHTNESS;
          }
          this.emit('brightness', settings.brightness);
        }
      },
      Preset: () => {
        settings.preset++;
        if (settings.preset === numPresets) {
          settings.preset = 0;
        }
        this.emit('preset', settings.preset);
      },
      Value1: () => this.handleValueChange(0, Direction.Up, steps),
      Value2: () => this.handleValueChange(1, Direction.Up, steps),
      Value3: () => this.handleValueChange(2, Direction.Up, steps),
      None: () => {}
    });
  }

  public controlDown(steps: number) {
    runOperationByControl(settings.currentControl, {
      Brightness: () => {
        if (settings.brightness > 0) {
          settings.brightness -= BRIGHTNESS_STEP * steps;
          if (settings.brightness < 0) {
            settings.brightness = 0;
          }
          this.emit('brightness', settings.brightness);
        }
      },
      Preset: () => {
        settings.preset--;
        if (settings.preset < 0) {
          settings.preset = numPresets - 1;
        }
        this.emit('preset', settings.preset);
      },
      Value1: () => this.handleValueChange(0, Direction.Down, steps),
      Value2: () => this.handleValueChange(1, Direction.Down, steps),
      Value3: () => this.handleValueChange(2, Direction.Down, steps),
      None: () => {}
    });
  }

  public clientConnected() {
    settings.numClients++;
  };

  public clientDisconnected() {
    settings.numClients--;
  };

  public getSettings() {
    return settings;
  }
}

export default new State();

File: coordinator/src/state.cpp
/*
Copyright (c) 2017 Bryan Hughes <bryan@nebri.us>

This file is part of Raver Lights.

Raver Lights is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Raver Lights is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Raver Lights.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <Arduino.h>
#include "state.h"
#include "common/codes.h"
#include "events.h"

const int BRIGHTNESS_STEP = 1;
const int MAX_BRIGHTNESS = 255;
const int VALUE_STEP = 1;
const int MAX_VALUE = 255;
const int DIM_TIMEOUT = 10000;
const int OFF_TIMEOUT = 5000;

bool isActive = false;
int idleCountdown = DIM_TIMEOUT;

State::Settings settings;

State::Settings* State::getSettings() {
  return &settings;
}

void State::nextControl() {
  int maxControls = 3;
  for (int i = 0; i < NUM_PRESET_VALUES; i++) {
    if (presetValueLabels[settings.preset][i] == NULL) {
      break;
    }
    maxControls++;
  }
  settings.currentControl++;
  if (settings.currentControl == maxControls) {
    settings.currentControl = 0;
  }

  Events::emitControlEvent(settings.currentControl);
}

int calculateNewValue(int value, bool direction) {
  if (direction) {
    value += VALUE_STEP;
    if (value > MAX_VALUE) {
      value = MAX_VALUE;
    }
    return value;
  } else {
    value -= VALUE_STEP;
    if (value < 0) {
      value = 0;
    }
    return value;
  }
}

void handleValueChange(int code, bool direction) {
  int newValue = calculateNewValue(settings.presetValues[settings.preset][code], direction);
  settings.presetValues[settings.preset][code] = newValue;
  Events::emitValueEvent(settings.preset, code, newValue);
}

void State::controlUp() {
  switch(settings.currentControl) {
    case Codes::Control::Brightness:
      if (settings.brightness < MAX_BRIGHTNESS) {
        settings.brightness += BRIGHTNESS_STEP;
        if (settings.brightness > MAX_BRIGHTNESS) {
          settings.brightness = MAX_BRIGHTNESS;
        }
        Events::emitBrightnessEvent(settings.brightness);
      }
      break;
    case Codes::Control::Preset:
      settings.preset++;
      if (settings.preset == NUM_PRESETS) {
        settings.preset = 0;
      }
      Events::emitPresetEvent(settings.preset);
      break;
    case Codes::Control::Value1:
      handleValueChange(0, true);
      break;
    case Codes::Control::Value2:
      handleValueChange(1, true);
      break;
    case Codes::Control::Value3:
      handleValueChange(2, true);
      break;
  }
}

void State::controlDown() {
  switch(settings.currentControl) {
    case Codes::Control::Brightness:
      if (settings.brightness > 0) {
        settings.brightness -= BRIGHTNESS_STEP;
        if (settings.brightness < 0) {
          settings.brightness = 0;
        }
        Events::emitBrightnessEvent(settings.brightness);
      }
      break;
    case Codes::Control::Preset:
      switch(settings.preset) {
        case Codes::Preset::Fade:
          settings.preset = Codes::Preset::Pulse;
          break;
        case Codes::Preset::Pulse:
          settings.preset = Codes::Preset::Fade;
          break;
      }
      Events::emitPresetEvent(settings.preset);
      break;
    case Codes::Control::Value1:
      handleValueChange(0, false);
      break;
    case Codes::Control::Value2:
      handleValueChange(1, false);
      break;
    case Codes::Control::Value3:
      handleValueChange(2, false);
      break;
  }
}

void State::setClientsConnected(int numClients) {
  if (numClients == settings.numClients) {
    return;
  }
  settings.numClients = numClients;
  Events::emitClientEvent(settings.numClients);
}

void State::setActive() {
  isActive = true;
  settings.idleState = Codes::IdleState::Active;
  Events::emitIdleEvent(settings.idleState);
}

void State::setIdling() {
  idleCountdown = DIM_TIMEOUT;
  isActive = false;
}

void State::init() {
  settings.presetValues = new byte*[NUM_PRESETS];
  for (int i = 0; i < NUM_PRESETS; i++) {
    settings.presetValues[i] = new byte[NUM_PRESET_VALUES];
    for (int j = 0; j < NUM_PRESET_VALUES; j++) {
      settings.presetValues[i][j] = presetValueDefaults[i][j];
    }
  }
  Serial.println("State initialized");
}

void State::loop() {
  if (!isActive && idleCountdown > 0) {
    idleCountdown--;
    if (!idleCountdown) {
      switch(settings.idleState) {
        case Codes::IdleState::Active:
          idleCountdown = OFF_TIMEOUT;
          settings.idleState = Codes::IdleState::ShallowIdle;
          Events::emitIdleEvent(settings.idleState);
          break;
        case Codes::IdleState::ShallowIdle:
          settings.idleState = Codes::IdleState::DeepIdle;
          Events::emitIdleEvent(settings.idleState);
          break;
      }
    }
  }
}
